### DISCUSSION TITLE
Replacing Chars

### DISCUSSION THREAD

**Discussion by fcabralj** [JoshClose/CsvHelper#1440] [Q&A]
In my IQueryable I have some illegal characters, for instance, '\n' and ';' which I would like to replace. However, in the way I did it I can't do it as I'm writting the query directly `csv.WriteRecords(query)`. How could I do that?

```
 public static void ConvertLINQToCsv<T>(string path, IQueryable<T> query)
  {
    using (var sw = new StreamWriter(path))
            using (var csv = new CsvWriter(sw))
            {
                csv.Configuration.Delimiter = UniversalVariables.csvDelimiter.ToString(); ;
                csv.Configuration.HasHeaderRecord = true;
                csv.Configuration.UseNewObjectForNullReferenceMembers = true;

                csv.WriteHeader<T>();
                csv.NextRecord();
                csv.WriteRecords(query);

                sw.Flush();
            }
  }

```

**Comment by JoshClose** [comment]
You would at least have to project the `IQueryable` to get the result. Othewise CsvHelper is going to map to the properties of the `IQueryable` class, and not the data it represents.

**Comment by fcabralj** [comment]
What do you mean by project the IQueryable? Do you mean, looping though it, replacing it? 
If that's correct how could I use csv helper without executing the query all over again?

**Comment by JoshClose** [comment]
Actually, what you're doing should work since `IQueryable` extends `IEnumerable`.

What do you mean by illegal characters? When writing fields that contain those characters, they'll be escaped.

Also, if you're doing `WriteRecords`, the header is automatically written for you. What you're doing will give you 2 header rows.

**Comment by fcabralj** [comment]
Thx for the header duplication tips, I would avoid using WriteHeader. 

Regarding the illegal chars, in one of the fields in "query" I have like this sentence:

"This is a sentence with \n illegal ;;;; chars".

I would just like to replace the \n and ;;;; by nothing. But if they are really escaped that's fine for me. Should I put something else in the csv.configuration or will  it do this anyway?

**Comment by JoshClose** [comment]
Have you tried it? The CSV format allow for delimiters and newlines to be contained within fields by escaping the fields. CsvHelper will do everything for you.

This is the default, so you can remove if i you want.

```cs
csv.Configuration.HasHeaderRecord = true;
```

When the code exits the `using` block, everything is flushed automatically, so you should be able to remove this line also.

```cs
sw.Flush();
```

**Comment by Imishin** [comment]
I faced a similar issue. I need to get rid of some symbols from each column in a row in a process of writing of csv string but I can't find such feature in `CsvHelper.CsvWriter.Configuration` nor `CsvHelper.CsvWriter.Context` or somewhere else.

**Comment by Imishin** [comment]
I tried to override method `ConvertUsing` of class `MemberMap<TClass, TMember>` and it didn't work correctly:

```
    public class CustomMemberMap<TClass, TMember> : MemberMap<TClass, TMember>
    {
        public CustomMemberMap(MemberInfo member) : base(member) { }

        public override MemberMap<TClass, TMember> ConvertUsing(Func<TClass, string> convertExpression) => base.ConvertUsing(row => convertExpression(row).Replace("\r", "").Replace("\n", ""));
    }

    public class CustomClassMap<TClass> : ClassMap<TClass> where TClass : new()
    {
        public override MemberMap<TClass, TMember> Map<TMember>(Expression<Func<TClass, TMember>> expression, bool useExistingMap = true) => new CustomMemberMap<TClass, TMember>(base.Map(expression, useExistingMap).Data.Member);
    }

    public sealed class CsvFileMapperAccommodation : CustomClassMap<Accommodation>
    {
        public CsvFileMapperAccommodation()
        {
            Map(m => m.Name).Name($"AccommodationName").ConvertUsing(x => $"{x.Name}");
            Map(m => m.ExternalKey).Name($"AccommodationExternalKey").ConvertUsing(x => $"{x.ExternalKey}");
            Map(m => m.Pension).Name($"AccommodationPension").ConvertUsing(x => $"{x.Pension}");
            Map(m => m.Category).Name($"AccommodationCategory").ConvertUsing(x => $"{x.Category}");
            Map(m => m.BuyingPrice).Name($"AccommodationBuyingPrice").ConvertUsing(x => $"{x.BuyingPrice}");
            Map(m => m.Margin).Name($"AccommodationMargin").ConvertUsing(x => $"{x.Margin}");
            Map(m => m.Price).Name($"AccommodationPrice").ConvertUsing(x => $"{x.Price}");
            Map(m => m.Quantity).Name($"AccommodationQuantity").ConvertUsing(x => $"{x.Quantity}");
            Map(m => m.Supplier).Name($"AccommodationSupplier").ConvertUsing(x => $"{x.Supplier}");
        }
    }
```

**Comment by JoshClose** [comment]
@Imishin If you need to remove characters, whey don't you do it before sending it through CsvHelper?

**Comment by Imishin** [comment]
@JoshClose Because I already have this data in an object, the object contains a lot of properties so in an output I have about 300 fields per row, therefore, it's too complicated and not elegant to change manually the object data in the beginning and the value for each of the fields in the end. So I only try to map this object into a csv string hoping that it's possible to change it on the fly.

**Comment by JoshClose** [comment]
You can use a custom type converter to do this.

```cs
void Main()
{
	var records = new List<Foo>
	{
		new Foo { Id = 1, Name = "one;two" },
		new Foo { Id = 2, Name = "three\nfour" },
	};
	using (var writer = new StringWriter())
	using (var csv = new CsvWriter(writer, CultureInfo.InvariantCulture))
	{
		csv.Configuration.RegisterClassMap<FooMap>();
		csv.WriteRecords(records);
		writer.ToString().Dump();
	}
}

public class Foo
{
	public int Id { get; set; }
	public string Name { get; set; }
}

public class FooMap : ClassMap<Foo>
{
	public FooMap()
	{
		Map(m => m.Id).TypeConverter<ReplaceConverter>();		
		Map(m => m.Name).TypeConverter<ReplaceConverter>();
	}
}

public class ReplaceConverter : DefaultTypeConverter
{
	public override string ConvertToString(object value, IWriterRow row, MemberMapData memberMapData)
	{
		var converter = row.Configuration.TypeConverterCache.GetConverter(memberMapData.Member.GetType());
		var s = converter.ConvertToString(value, row, memberMapData);
		s = Regex.Replace(s, "[\n;]", string.Empty);
		
		return s;
	}
}
```

**Reply by christianwright1974** [reply]
i needed to format some fields with padded leading zeros. this helped me out so thank for that

**Comment by Imishin** [comment]
I got it, but I don't wanna replicate the applying of this converter for each field/property of the class: 
```
Map(m => m.Id).TypeConverter<ReplaceConverter>();		
Map(m => m.Name).TypeConverter<ReplaceConverter>();
```
Because in a case of multiple fields/properties I have to apply "copy-paste" practice :(
And besides I wanna be able to apply my own specific convertion for the particular field/property through `.ConvertUsing(x => ...)` don't loosing already mentioned custom type converter.

**Comment by Imishin** [comment]
Could you help me to override the method `ConvertUsing`, please? According to my opinion, it will be the best solution regarding this case.

**Comment by JoshClose** [comment]
You do the same thing as you would in a type converter. `ConvertUsing` is a way to inline convert in the map. You'll need to figure out how you want to convert your property to a string though.

```cs
public class FooMap : ClassMap<Foo>
{
	public FooMap()
	{
		Map(m => m.Id).ConvertUsing((field) =>
		{
			// Convert to string however you like. This is using
			// .NET's built in type conversion system.
			var converter = TypeDescriptor.GetConverter(typeof(int));
			var s = converter.ConvertToString(field.Id);
			s = Regex.Replace(s, "[\n;]", string.Empty);
			return s;
		});
		Map(m => m.Name).ConvertUsing((Foo field) =>
		{
			// Convert to string however you like. This is using
			// .NET's built in type conversion system.
			var converter = TypeDescriptor.GetConverter(typeof(string));
			var s = converter.ConvertToString(field.Name);
			s = Regex.Replace(s, "[\n;]", string.Empty);
			return s;
		});
	}
}
```

**Comment by 9Rune5** [comment]
I fear I'm in the same boat as @Imishin ...

I have a dozen or so (and counting...) classes that are exported. They all descend from a generic `CommonExport<TOption, TModel>` class. Inside CommonExport I set up the export:

                using var csv = new CsvWriter(writer);
                csv.Configuration.Delimiter = "\t";
                csv.Configuration.ShouldQuote = (s, context) => false;
                csv.WriteHeader<TModel>();

Now... I have been instructed to not quote any of the fields. And my counterpart doesn't expect newlines in the non-quotable string fields. 

A dynamic map has served me very well so far. It makes it very easy to add more DTOs in the future. Assigning a converter to each field looks awkward.

Can I have a global converter for the string type?

**Comment by JoshClose** [comment]
This will replace the current `StringConverter` globally with your own `MyStringConverter`.

```cs
csv.Configuration.TypeConverterCache.AddConverter<string>(new MyStringConverter());
```

**Comment by 9Rune5** [comment]
Ah, I struggled a bit until I noticed ConvertToString() was the one I was supposed to override.

Thanks!

**Comment by davidcassell** [comment]
Hi Josh, I have the same thing here. How do use the globally replaced converter? I add to my config...config.TypeConverterCache.AddConverter<string>(new RemoveControlCharacterStringConverter()); but what change do I need to do in my class mapping? Thanks David

**Comment by JoshClose** [comment]
@davidcassell You shouldn't have to make a change in your mapping. If you're specifying a type converter on a property in your mapping though, then that type convertible will take precedence. Since you're adding it as a converter for a `string` type, it will only work on properties whose type is a `string`.

**Comment by AltruCoder** [comment]
@davidcassell One thing I've found is the Converter has to be added to the TypeConverterCache before the mapping is added.

**Comment by davidcassell** [comment]
Thanks guys. it works well
